segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/TruncateOperation.java

```diff
@@ -133,9 +139,98 @@
+                        currentMetadata.setActive(false);
+
+                        // Add block index entries.
+                        chunkedSegmentStorage.addBlockIndexEntriesForChunk(txn,
+                                segmentMetadata.getName(),
+                                newChunkName,
+                                segmentMetadata.getFirstChunkStartOffset(),
+                                segmentMetadata.getFirstChunkStartOffset(),
+                                segmentMetadata.getFirstChunkStartOffset() + newFirstChunkMetadata.getLength()
+                                );
+
+                        isFirstChunkRelocated = true;
+                        currentMetadata = newFirstChunkMetadata;
+                        currentChunkName = newChunkName;
+                    }, chunkedSegmentStorage.getExecutor());
+        }
+        return CompletableFuture.completedFuture(null);
+    }
+
+    private CompletableFuture<Void> copyBytes(ChunkHandle writeHandle, ChunkHandle readHandle, long startOffset, int length) {
+        val bytesToRead = new AtomicLong(length);
+        val readAtOffset = new AtomicLong(startOffset);
+        val writeAtOffset = new AtomicLong(0);
+        return Futures.loop(
+                () -> bytesToRead.get() > 0,
+                () -> {
+                    val buffer = new byte[Math.toIntExact(Math.min(chunkedSegmentStorage.getConfig().getMaxBufferSizeForChunkDataTransfer(), bytesToRead.get()))];
+                    return chunkedSegmentStorage.getChunkStorage().read(readHandle, readAtOffset.get(), buffer.length, buffer, 0)
+                            .thenComposeAsync(size -> {
+                                bytesToRead.addAndGet(-size);
+                                readAtOffset.addAndGet(size);
+                                return chunkedSegmentStorage.getChunkStorage().write(writeHandle, writeAtOffset.get(), size, new ByteArrayInputStream(buffer, 0, size))
+                                        .thenAcceptAsync(writeAtOffset::addAndGet, chunkedSegmentStorage.getExecutor());
+                            }, chunkedSegmentStorage.getExecutor());
+                },
+                chunkedSegmentStorage.getExecutor()
+        );
+    }
+
+    private boolean shouldRelocate() {
+        return chunkedSegmentStorage.getConfig().isRelocateOnTruncateEnabled()
+            && chunkedSegmentStorage.shouldAppend()
+            && !chunkedSegmentStorage.isSegmentInSystemScope(handle)
+            && currentMetadata.getLength() >  chunkedSegmentStorage.getConfig().getMinSizeForTruncateRelocationInbytes()
+            && getWastedSpacePercentage() >= chunkedSegmentStorage.getConfig().getMinPercentForTruncateRelocation();
+    }
+
+    private long getWastedSpacePercentage() {
+        val wastedSpace = segmentMetadata.getStartOffset() - segmentMetadata.getFirstChunkStartOffset();
+        return 100 * wastedSpace / currentMetadata.getLength();
```

Comment:
can there be a division by zero here?

Reply:
As it happens when currentMetadata.length==0 the entire chunk is deleted and then this method is not called.
But I'll fix this method to remove possibility of divide by 0.

Reply:
Fixed.